### ISSUE TITLE
Remove the @open_file decorator

### ISSUE THREAD

**Issue by jfinkels** [networkx/networkx#2298]
In working on a fix for issue #2295, I found the `@open_file` decorator confusing. Correct me if I'm wrong, but I think it's purpose is to allow the possibility of, for example, both `write_graph6(G, '/tmp/foo')` *and* `with open('/tmp/foo', 'w') as f: write_graph6(G, f)`. Modern versions of Python make opening files (and creating file-like objects like `StringIO`) easy, so this layer of abstraction seems unnecessary.

I propose removing the `open_file` decorator and changing the signature of functions that allow a path as input to require an (open) file-like object instead. This would require users to replace code like this:
```python
write_graph6(G, '/tmp/foo')
```
with code like this:
```python
with open('/tmp/foo', 'w') as f:
    write_graph6(G, f)
```

**Comment by chebee7i** [comment]
It could be nice to remove it. Are you proposing that the NetworkX API will only accept file objects, and that strings will no longer be allowed as input? You've correctly identified part of the reason for `@open_file`---and a simpler API would mean that it's not necessary.

This seems like a pretty big API change though, right?  All functions will only accept file objects, and now NetworkX will never close them.  Right now, we close file objects that we create, but do not touch file objects that are passed in.

**Comment by chebee7i** [comment]
To be clear, I'm not opposed to making the change...just wanted to clarify the ramifications.

**Comment by jfinkels** [comment]
Yes, I am proposing that the read/write functions no longer accept a string naming the path to a file, and all read/write functions deal only in file-like objects. This is a major API change in the sense that many functions will change and it is not backwards-compatible. Although
```python
with open('/path/to/file') as f:
    networkx.write_whatever(f)
```
forces some boilerplate code on the user, the pattern is ubiquitous in Python, clear, explicit, and only one extra line of code. The benefit for NetworkX is that it not only eliminates a function that I found confusing (a selfish reason :) ), but also removes the complexity of file management. The counterargument is that NetworkX should be "batteries included": a user should be able to just write `networkx.write_whatever('/path/to/file')` to write to a file. I guess whether removing this removes a battery is up to you, I don't really think so.

**Comment by rossbar** [comment]
> Yes, I am proposing that the read/write functions no longer accept a string naming the path to a file, and all read/write functions deal only in file-like objects.

I'm -1 on this change: having a hard-to-grok decorator doesn't seem to me like a strong motivation for removing a very convenient, user-facing feature. I think it's also a pretty well-established pattern (at least in the scientific Python ecosystem AFAIK) for file IO functions to handle the actual file management for the user (see e.g. the load/save functions in NumPy).

Re: the decorator itself, I too found it difficult to read (a quick look at `git blame` shows that it was introduced ~9 yrs ago and, aside from formatting, hasn't really been touched since then). I'm not at all opposed to removing it and/or refactoring how/where file handling is done.

To sum up: I'm +1 on trying to remove/improve `open_file`, but -1 on any changes that remove the "convenience" support for `str`, `pathlib.Path`, etc.

**Comment by MridulS** [comment]
It looks like we decided on **not** removing the convenience of the `@open_file` decorator. I'll go ahead and close this but feel free to reopen an issue if someone wants to start the conversation again :)
